### ISSUE TITLE
commitasync triggers assert, one  opaque value remains 0, 1 proposal and a few questions 

### ISSUE THREAD

**Issue by FDePourcq** [edenhill/librdkafka#528]
Hi Magnus, 

After a week of Kafka 0.9 I came up with the following questions  & issues for you, sorry for the spam :)

1. How to specify the desired offset with kafkaconsumer? setting offsets in the rebalance-callback doesnt seem to have any effect.

2. The per-message delivery report callback set with rd_kafka_conf_set_dr_msg_cb is supposed to receive a opaque value passed to rd_kafka_produce, 
   this opaque value remains 0.

3. The rebalancing with kafkaconsumer, to me it seems that kafka iterates every consumer one by one and calls the rebalancecallback with the list of not-yet-assigned partitions.
   This list shrinks with every call. Is this the correct interpretation? 
   I have seen a partitionlist of size 1 for a topic with 2 partitions and that is how I came to the above interpretation.
   The whole rebalancing isn't really clear to me. 
   I wonder for example if it really is a bad thing if a consumer refuses to unassign his partitions when the revoke happens.

4. The kafkaconsumer-interface forces me to either commit all received messages at the same time or to keep the old messages in memory
   till they are really processed. Depending on the application this could be many messages later, when a expected sequence of messages has been completed.
   The commit(A)sync(message)-code only seems to need the topic+partition+offset and not the message. 
   For that reason i propose to extend the kafkaconsumer-api with commit(A)Sync(topic,const uint32_t partition,const uint64_t offset)

5. When calling KafkaConsumer::commitAsync (both the version with message and the one without) an assert triggers for me at every run.
   rdkafka_cgrp.c: rd_kafka_cgrp_handle_OffsetCommit 
	rd_kafka_assert(rkcg->rkcg_rk,
                        rkcg->rkcg_flags & RD_KAFKA_CGRP_F_WAIT_COMMIT);
   Removing the assert gives me a neatly running program but I need to inquire the great wizard about the gravity of this assert.

**Comment by edenhill** [comment]
Hey,

thanks for the detailed list of issues, very helpful.

I'll start with the easy one:

 **2.** The per-msg opaque is actually in rd_kafka_message_t._private, the opaque argument to dr_msg_cb is the global opaque (rd_kafka_conf_set_opaque).

 **4.** Ah, yeah, that's what the C API looks like: https://github.com/edenhill/librdkafka/blob/master/src/rdkafka.h#L1766
   I'll make sure there is a corresponding one for the C++ API.

**Comment by FDePourcq** [comment]
Hi thanks for the answer, 
About the old issues and adding new issues:
1. still an issue, seems to be a duplicate of " Manually controlling offsets in rebalance callbacks #458 "
2. solved, global opaque values work for me
3. postpone
4. thanks, i see you will do the main issues first, the C++ api can wait, kafkacat shows how to do it in C.
5. The assert doesnt trigger anymore for me with the latest revisions (both kafka+rdkafka) for the kafkacat-inspired c-kafkaconsumer. 

And new ones:
6. in the statistics-callback-json, what is partition -1 ?? (it seems not to contain anything of interest, useless code hickup?)
7. in the statistics-callback-json, there is a commited_offset as well as a committed_offset, just to be sure you know.

8. when trying to not advance the stored offset by disabling autocommit and not calling rd_kafka_offset_store:
 - The value from statistics callback partition0_stored_offset still increases.
 - If statistics callback partition0_stored_offset in the previous run reported a certain offset X then the next run could still start consuming before X. But the run after that one really starts with X.
  Basically, it seems that this usecase whereby consumers want to delay the advancing of the offset, isn't really there yet, maybe not supported even?

**Comment by edenhill** [comment]
Quick:
 6. partition -1, also called the Unassigned partition or PARTITION_UA, is an internal object that temporary holds messages waiting to be partitioned, waiting for the proper partition, etc. It makes sense to include it in stats for troubleshooting purposes since it can hold quite a large number of messages.
 7. they are both the same, the misspelled one is still there due to legacy reasons (must not break old apps).

**Comment by FDePourcq** [comment]
More on 8. 
Your aliases for autocommit in the global conf were messing me up:
% ERROR: No such configuration property: "auto.commit.enable"

But "enable.auto.commit" works... maybe that is worth another bug.
while for the the topic config i assume both work but do something else.

--> This way i managed to avoid that rd_kafka_cgrp_offset_commit_tmr_cb gets called which would have enqueued the offsetcommit on its way to the broker.
Though, ... i now have the opposite issue: with the autocommit finally disabled I see that rd_kafka_offset_store doesn't really trigger a commit of the offset to the broker at all.
 
*continues digging*

**Comment by edenhill** [comment]
`auto.commit.enable` is the legacy consumer *topic* configuration property, it has no affect when using the high-level KafkaConsumer which uses the *global* `enable.auto.commit`.
This unfortunate naming inconsistency is due to following the upstream Kafka clients naming.

**Comment by edenhill** [comment]
rd_kafka_offset_store() only **stores** the offset for future commit, you will still need to commit the stored offsets every now and then with one of the commit APIs.

**Comment by FDePourcq** [comment]
rd_kafka_commit does its job ... but now i'm thinking ... I want a batch of offsetcommits to be reduced to just the highest offset every x time units, i want a timer, i want only one offsetcommit every interval ... but i want the stored offset to be committed, not the offset of the latest message that was handed to me, it has to be that highest offset of which i explicitly stated that it could be stored.
(while now the autocommit ... just commits whatever offset it handed out without regard for the opinion of the user)

Did I miss something really basic here?

**Comment by edenhill** [comment]
What you are describing is auto-commit, unfortunately (for you) the auto-commit feature also does auto-store for each message provided to your app, so that wont work for you.

I guess we could add a `enable.auto.offset.store=false` that keeps the auto-commit but disables the auto-storing.

Sounds ok?

**Comment by FDePourcq** [comment]
yes yes yes :)

**Comment by edenhill** [comment]
Okay, I'll have a patch for you later tonight.

**Comment by FDePourcq** [comment]
w00t :) :+1:

**Comment by FDePourcq** [comment]
(ow btw, the triggering assert from issue 5 during async commit is still there sorry)

**Comment by edenhill** [comment]
Yepyep, I'm working on it.

**Comment by edenhill** [comment]
*3.* The rebalancing starts off with the broker electing a leader client, that leader performs the actual assignment calculation using the negotiated assignment strategy  (e.g, range or roundrobin). When it is done it sends the complete assignment back to the broker. The broker sends each client's its new assignment, it will do so simultaneously.

A client may choose not to use its assignment, but the only way to "reject" it would be to be leave the consumer group. I would say this use-case is very atypical and contradicts the purpose of consumer groups. (iow: Dont do this)

There is a bug in librdkafka's assignment strategy when different clients have different subscription sets, this is also considered atypical.

**Comment by FDePourcq** [comment]
Ah, I heard that the partition-to-consumer-assignment-strategy was clientside and thought for a while it had to be in the assignment callback, I was wrong. There doesnt seem to be a API to modify that algorithm at the moment ...  But ... if i can guarantee that all consumers are my hacked consumers ... I could modify/add the strategy within rdkafka ... to make partion-consumer-combinations sticky ...

**Comment by edenhill** [comment]
The plan is to expose an assignment strategy API eventually, but go ahead and hack it in the meantime and let me know if you find any short-comings (that would prove useful in the future API).

**Comment by FDePourcq** [comment]
jeej :)  *opens pandoras box*

**Comment by edenhill** [comment]
Update:
 1. fixed
 2. answered
 3. answered
 4: fixed
 5: fixed
 6: answered
 7: answered
 8: fixed

Please verify on your end

**Comment by FDePourcq** [comment]
Thanks! Merged it and things seem to work fine.

Fyi, I have a hack here that allows the user to register a callback for partition-assignment ... supports just one topic though. (My sticky rebalancing code is all c++ stl, so no good within rdkafka.)

[addedrdkafka_callback_assignor.diff.zip](https://github.com/edenhill/librdkafka/files/126571/addedrdkafka_callback_assignor.diff.zip)
